File: src/executionenvironments/r/pcalg.r
library(optparse, quietly = T)
library(pcalg, quietly = T)
source("/scripts/mpci_utils.r")

option_list_v <- list(
                    make_option(c("-j", "--job_id"), type="character",
                                help="Job ID", metavar=""),
                    make_option(c("--api_host"), type="character",
                                help="API Host/Port", metavar=""),
                    make_option(c("-d", "--dataset_id"), type="character",
                                help="Dataset ID", metavar=""),
                    make_option(c("-t", "--independence_test"), type="character", default="gaussCI",
                                help="Independence test used for the pcalg", metavar=""),
                    make_option(c("-a", "--alpha"), type="double", default=0.05,
                                help="This is a hyperparameter", metavar=""),
                    make_option(c("-c", "--cores"), type="integer", default=1,
                                help="The number of cores to run the pc-algorithm on", metavar=""),
                    make_option(c("-s", "--subset_size"), type="integer", default=-1,
                                help="The maximal size of the conditioning sets that are considered", metavar=""),
                    make_option(c("--send_sepsets"), type="integer", default=0,
                                help="If 1, sepsets will be sent with the results", metavar=""),
                    make_option(c("-v", "--verbose"), type="integer", default=0,
                                help="More detailed output is provided (with impact on performance)", metavar="")
                    #make_option(c("--fixed_gaps"), type="character", default=NULL,
                    #            help="The connections that are removed via prior knowledge", metavar=""),
                    #make_option(c("--fixed_edges"), type="character", default=NULL,
                    #            help="The connections that are fixed via prior knowledge", metavar=""),

);

indepTestDict <- list(gaussCI=gaussCItest, binCI=binCItest, disCI=disCItest)

option_parser <- OptionParser(option_list=option_list_v)
opt <- parse_args(option_parser)

df <- get_dataset(opt$api_host, opt$dataset_id, opt$job_id)

matrix_df <- data.matrix(df)

if (opt$independence_test == "gaussCI") {
    sufficient_stats <- list(C=cor(matrix_df), n=nrow(matrix_df))
} else if (opt$independence_test == "binCI") {
    sufficient_stats <- list(dm=matrix_df, adaptDF=FALSE)
} else if (opt$independence_test == "disCI"){
    p <- ncol(matrix_df)
    nlev <- vapply(seq_len(p), function(j) length(levels(factor(matrix_df[,j]))), 1L)
    sufficient_stats <- list(dm=matrix_df, adaptDF=FALSE, nlev=nlev)
} else{
    stop("No valid independence test specified")
}

subset_size <- if(opt$subset_size < 0) Inf else opt$subset_size
verbose <- opt$verbose > 0
result = pc(suffStat=sufficient_stats, verbose=verbose,
            indepTest=indepTestDict[[opt$independence_test]], m.max=subset_size,
            p=ncol(matrix_df), alpha=opt$alpha, numCores=opt$cores, skel.method="stable.fast")

graph_request <- store_graph_result(opt$api_host, result@'graph', result@'sepset', df, opt$job_id, opt$independence_test, opt$send_sepsets, opt)

File: services/executionenvironments/python/pcalg.py
import argparse
import logging
import time

from cmi_estimators import GKOVEstimator, CCMIEstimator
from mpci_utils import get_dataset, store_graph_result
from py_parallel_pc import parallel_stable_pc
from rperm import RPermTest

parser = argparse.ArgumentParser(description='Parse PCAlg parameters.')
parser.add_argument('-j', '--job_id', type=str, help='Job ID')
parser.add_argument('--api_host', type=str, help='API Host/Port')
parser.add_argument('-d', '--dataset_id', type=str, help='Dataset ID')
parser.add_argument('-t', '--independence_test', choices=['RPerm_GKOV', 'RPerm_CCMI'],
                    type=str, help='Independence test used for the pcalg', default='RPerm_GKOV')
parser.add_argument('-a', '--alpha', type=float, help='The significance level of the test', default=0.05)
parser.add_argument('-p', '--processes', type=int, default=1,
                    help='The number of separate processes for parallelization')
parser.add_argument('-s', '--subset_size', type=int,
                    help='The maximal size of the conditioning sets that are considered')
parser.add_argument('--permutations', type=int, default=None,
                    help='The number of iterations for the permutation test')
parser.add_argument('--send_sepsets', type=int, default=0, help='If 1, send sepsets with the results')


def run_pcalg(api_host, job_id, dataset_id, indep_test, alpha, processes, max_level, args, send_sepsets=False):
    df, ds_load_time = get_dataset(api_host, dataset_id, job_id)

    start = time.time()
    graph, sepsets = parallel_stable_pc(df, indep_test, alpha=alpha, processes=processes, max_level=max_level)
    exec_time = time.time() - start

    store_graph_result(api_host, job_id, graph, exec_time, ds_load_time, args,
                       sepsets=(sepsets if send_sepsets else None))


if __name__ == '__main__':
    logging.getLogger().setLevel(logging.INFO)
    args = parser.parse_args()

    indep_tests = {
        'RPerm_GKOV': lambda: RPermTest(GKOVEstimator(k='adaptive'), k=15,
                                        iterations=args.permutations if args.permutations is not None else 100),
        'RPerm_CCMI': lambda: RPermTest(CCMIEstimator(), k=15,
                                        iterations=args.permutations if args.permutations is not None else 10),
    }
    indep_test = indep_tests[args.independence_test]()
    send_sepsets = args.send_sepsets == 1
    max_level = args.subset_size if args.subset_size >= 0 else None

    run_pcalg(args.api_host, args.job_id, args.dataset_id, indep_test, alpha=args.alpha,
              processes=args.processes, max_level=max_level, args=vars(args), send_sepsets=send_sepsets)
